feat: Ontologies to semantic

RedingJeff opened the pull request:
Finished the work on the semantic addon for Filip.

Filip is now able to parse and combine ontologies into a vocabulary.
After the vocabulary is defined it can converted to Python models and semantic instances will be instantiated and linked by using Python objects.
The Filip functionalities are used in the background to pull the needed state information from Fiware (in Hot-loading) and push the local edited state into Fiware while ensuring a defined state and minimal information loss.

tstorek commented on filip/semantics/vocabulary/entity.py:
```diff
@@ -0,0 +1,164 @@
+    label
+    """
+    iri: str
+    """Unique Internationalized Resource Identifier """
```
pleas use pydantic fields

tstorek commented on filip/semantics/vocabulary/entity.py:
```diff
@@ -0,0 +1,164 @@
+
+
+unallowed_labels = ["id","type","geo:property", "metadata"]
+"""Labels that are not allowed, as the would lead to errors in FIWARE"""
```
inline commets with #

tstorek commented on filip/semantics/vocabulary/entity.py:
```diff
@@ -0,0 +1,164 @@
+    from . import Vocabulary
+
+
+unallowed_labels = ["id","type","geo:property", "metadata"]
```
missing whitespace. you should probably check pylint in the CI

tstorek commented on filip/semantics/semantic_models.py:
```diff
@@ -0,0 +1,1683 @@
+import uuid
```
modukle docstring

tstorek commented on filip/semantics/semantic_models.py:
```diff
@@ -0,0 +1,1683 @@
+    DataFields
+    """
+    type: str
+    """Type of the datatype"""
```
Fields?

tstorek commented on filip/semantics/semantic_models.py:
```diff
@@ -0,0 +1,1683 @@
+
+            return True
+
+        if self.type == "enum":
```
is this always lowercase?

tstorek replied:
stiil open?

tstorek commented on filip/semantics/semantic_models.py:
```diff
@@ -0,0 +1,1683 @@
+    multiple fields will result in an error.
+    """
+
+    name: str
```
Field?

tstorek commented on filip/semantics/semantic_models.py:
```diff
@@ -0,0 +1,1683 @@
+        underscore_attrs_are_private = True
+
+
+class Command(DeviceProperty):
```
This seems to me confusing since there is alerady a class for commands

tstorek commented on filip/semantics/semantic_models.py:
```diff
@@ -0,0 +1,1683 @@
+    """
+    Retrieval type of the DeviceAttribute value from the IoT Device into Fiware
+    """
+    lazy = "lazy"
```
please use anootated enums here

tstorek commented on filip/semantics/semantic_models.py:
```diff
@@ -0,0 +1,1683 @@
+    """The value is kept up-to-date"""
+
+
+class DeviceAttribute(DeviceProperty):
```
why issn#t this mdoel in iot?

RedingJeff replied:
It is a higher order abstraction of the object in IoT models.

tstorek commented on filip/semantics/semantic_models.py:
```diff
@@ -0,0 +1,1683 @@
+        use_enum_values = True
+
+
+class Field(BaseModel):
```
this might conflict with pydantic Field

RedingJeff replied:
Yes, but only at one point i needed to use pydantic.Fiedl ot avoid conflict. As normaly this Field is not imported in other modules. And the name describes the best what it is

tstorek commented on filip/semantics/vocabulary/datatype.py:
```diff
@@ -0,0 +1,136 @@
+from . import Entity
+
+
+class DatatypeType(str, Enum):
```
can't we reuse somehow?

RedingJeff replied:
Reuuse from what? I think they are two fundamentaly different from any other existing datatype enums in the project

tstorek commented on filip/semantics/vocabulary/datatype.py:
```diff
@@ -0,0 +1,136 @@
+    """If Type==String: Blacklisted chars"""
+    allowed_chars: List[str] = []
+    """If Type==String: Whitelisted chars"""
+    enum_values: List[str] = []
```
Pydantic Fieldsß

tstorek commented on filip/semantics/vocabulary/object_property.py:
```diff
@@ -0,0 +1,54 @@
+from typing import List, Set, TYPE_CHECKING
```
moduke docstrings

tstorek commented on filip/semantics/vocabulary/object_property.py:
```diff
@@ -0,0 +1,54 @@
+    def is_logically_equivalent_to(self, object_property: 'ObjectProperty',
+                                   vocabulary: 'Vocabulary',
+                                   old_vocabulary: 'Vocabulary') -> bool:
+
```
remove empty line

tstorek commented on filip/semantics/vocabulary/object_property.py:
```diff
@@ -0,0 +1,54 @@
+    """
+
+    inverse_property_iris: Set[str] = set()
+    """List of property iris that are inverse:Of; 
```
Fields

tstorek commented on filip/semantics/vocabulary/relation.py:
```diff
@@ -0,0 +1,276 @@
+from typing import List, Set, Dict, TYPE_CHECKING
```
Module dcostrings

tstorek commented on filip/semantics/vocabulary/relation.py:
```diff
@@ -0,0 +1,276 @@
+
+
+class RestrictionType(str, Enum):
+    """RestrictionTypes, as defined for OWL:
```
start second line. First it short descriptiOn

tstorek commented on filip/semantics/vocabulary/relation.py:
```diff
@@ -0,0 +1,276 @@
+        exactly: exactly n values of that target
+        value: predefined value
+    """
+    some = 'some'
```
you can also use annotated enums here

tstorek commented on filip/semantics/vocabulary/relation.py:
```diff
@@ -0,0 +1,276 @@
+    value = 'value'
+
+
+class Relation(BaseModel):
```
Realtion vs Realtionship?

tstorek replied:
Hence, we stick with relation?

tstorek commented on filip/semantics/vocabulary/relation.py:
```diff
@@ -0,0 +1,276 @@
+    """
+
+    id: str
+    """Unique generated Relation ID, for internal use"""
```
Fields?

tstorek commented on filip/semantics/vocabulary/relation.py:
```diff
@@ -0,0 +1,276 @@
+        if self.restriction_type == RestrictionType.value:
+            return number_of_fulfilling_values >= 1
+
+
```
to many blank lines

tstorek commented on filip/semantics/vocabulary/relation.py:
```diff
@@ -0,0 +1,276 @@
+                   + self.target_statement.to_string(vocabulary)
+
+    def is_restriction_fulfilled(self, number_of_fulfilling_values: int,
+                                 total_number_of_values: int) -> bool:
```
ssome of the function coudl be validators right?

RedingJeff replied:
I would say no. As this validation and pydantic validators work on tow different levels. 
A pydantic validator checks if the object is allowed to exist with given values, this validator operates one level higher on semantic reasoning.

Further the methode is not used in the object itself

tstorek commented on filip/semantics/vocabulary/source.py:
```diff
@@ -0,0 +1,187 @@
+    A dependency is a reference of one iri in an other entity definition
+    """
+    source_iri: str = ""
+    """Iri of the source containing the statement"""
```
use fields?

tstorek commented on filip/semantics/vocabulary/target_statment.py:
```diff
@@ -0,0 +1,267 @@
+from enum import Enum
```
typo in filename. missing module docstring

tstorek commented on filip/semantics/vocabulary/target_statment.py:
```diff
@@ -0,0 +1,267 @@
+                    lengths.append(len(sublist))
+
+                # init with empty lists
+                for c in range(number_of_entries):
```
be more pythonic. a variable should have at least 3 characters

RedingJeff replied:
c was used here as name, as the variable is never used in the for loop. And to my knowledge python does not offer a cleaner way to loop a range without variable. But I changed it now to empty

tstorek commented on filip/semantics/vocabulary/target_statment.py:
```diff
@@ -0,0 +1,267 @@
+
+    def get_dependency_statements(
+            self,
+            vocabulary: 'Vocabulary',
```
THis must be equal to this string?

RedingJeff replied:
Equal to which string, could you elaborate?

tstorek commented on filip/utils/validators.py:
```diff
@@ -69,10 +70,10 @@ def validate_escape_character_free(value: Any) -> Any:
-                                  f"the forbidden char \"")
-            if "'" in v:
+                                 f"the forbidden char \"")
+            if "'" == v[-1:] or "'" == v[0:1]:
```
pelas substitute v with somehting more descriptive

tstorek commented on tests/.env:
```diff
@@ -0,0 +1,7 @@
+LOG_LEVEL="info"
+CB_URL="http://localhost:1026"
+IOTA_URL="http://localhost:4041"
```
This is not valid for the CI anymore

tstorek commented on tests/clients/test_ngsi_v2_iota.py:
```diff
@@ -109,7 +109,7 @@ def test_device_endpoints(self):
                                             entity_name='pressure')
             attr_static = StaticDeviceAttribute(name='hasRoom',
-                                                type='Relationship',
+                                                type='RelationField',
```
Does relationshop not exist anymore?

RedingJeff replied:
Refactor overshot, Changed back

tstorek commented on examples/semantic/models.py:
```diff
@@ -0,0 +1,432 @@
+from enum import Enum
+from typing import Dict, Union, List
+from filip.semantics.semantic_models import \
+	SemanticClass, SemanticIndividual, RelationField, DataField, SemanticDeviceClass, DeviceAttributeField,CommandField
```
The package should be named "semantics"

tstorek commented on tests/semantic/.env.filip:
```diff
@@ -0,0 +1,3 @@
+CB_URL="http://localhost:1026"
```
please move tests to tests/semantics

RedingJeff replied:
File should not exist anymore; now deleted

tstorek commented on filip/semantics/vocabulary/vocabulary.py:
```diff
@@ -0,0 +1,493 @@
+
+
+class IdType(str, Enum):
+    class_ = 'Class'
```
docstring or aenum

tstorek replied:
This is still open. The documentation is missing

tstorek commented on filip/semantics/vocabulary/vocabulary.py:
```diff
@@ -0,0 +1,493 @@
+
+
+class LoggingLevel(str, Enum):
+    """LoggingLevel for parsing statements"""
```
why not use standard loglevels hereß

RedingJeff replied:
These are not for output on the console, but create separate entries that are saved and can later be viewed by the user when inspecting the vocabulary. Therefore the separation from the classical LoggingLevel

tstorek replied:
please double check this LogLevel Enum, maybe we should integrate it somewhere 

tests/config.py

tstorek commented on filip/semantics/vocabulary/vocabulary.py:
```diff
@@ -0,0 +1,493 @@
+
+
+class VocabularySettings(BaseModel):
+    pascal_case_class_labels: bool = True
```
fields ? or docstrings missing

tstorek commented on filip/semantics/vocabulary/vocabulary.py:
```diff
@@ -0,0 +1,493 @@
+        return list(self.classes.values())
+
+    def get_classes_sorted_by_label(self) -> List[Class]:
+        return sorted(self.classes.values(),
```
many missing docstrings

tstorek commented on filip/semantics/vocabulary_configurator.py:
```diff
@@ -0,0 +1,769 @@
+import copy
```
docstring

tstorek commented on filip/semantics/vocabulary_configurator.py:
```diff
@@ -0,0 +1,769 @@
+    Model holding all information for label conflicts in a vocabulary
+    """
+    class_label_duplicates: Dict[str, List[Entity]]
+    """All Labels that are used more than once for class_names on export. 
```
Fields?

tstorek reviewed (changes requested):
Respact. THis is a lot of stuff :) 👍

RedingJeff commented on examples/semantic/semantics_model_example.py:
```diff
@@ -0,0 +1,471 @@
+    # given:
+    my_building.goalTemperature.add(23)
+    my_building.goalTemperature.update([12])
+    my_building.goalTemperature.remove(23)
```
Remove is the standard remove of a set(). 
It removes the given value from the Field.

RedingJeff commented on examples/semantic/semantics_model_example.py:
```diff
@@ -0,0 +1,471 @@
+    # To manipulate data inside a field all standard set interactions are
+    # given:
+    my_building.goalTemperature.add(23)
+    my_building.goalTemperature.update([12])
```
Update is build as the update of a set().
It takes a list or a set and adds them to the Field values.

RedingJeff commented on examples/semantic/semantics_vocabulary_example.py:
```diff
@@ -0,0 +1,256 @@
+    for source in vocabulary.get_source_list():
+        print(f'Name: {source.source_name}; Added: {source.timestamp}; '
+              f'Id: {source.id}')
+    print()
```
An output for the user who looks at the example, to visualise him the internal state. It produces this output:

S̲o̲u̲r̲c̲e̲s̲ ̲i̲n̲ ̲v̲o̲c̲a̲b̲u̲l̲a̲r̲y̲:
Name: building circuits; Added: 2021-11-22 11:13:36.235140; Id: a012f8f1a36b4d2fa4dee5f6fc5c5aa6
Name: ParsingExample; Added: 2021-11-22 11:13:36.259138; Id: 3579d3cce7584b338f64cac8f074d8dc
Name: saref.ttl; Added: 2021-11-22 11:13:36.505593; Id: 5aebcfca18fd452f9e19cdb394110cfe
Name: Predefined; Added: 2021-11-22 11:13:36.632298; Id: PREDEFINED

tstorek replied:
I mean the empty print

RedingJeff commented on examples/semantic/semantics_vocabulary_example.py:
```diff
@@ -0,0 +1,256 @@
+    # the parsing logs:
+    print("\u0332".join("Parsing Logs of vocabulary:"))
+    print(VocabularyConfigurator.get_parsing_logs(vocabulary))
+    print()
```
I decided here for the empty print, as I else would need to add an f'{}'\n to the above line, reaching therefore more than 80 chars and having an ugly line break.
It is more readable in this notation in my opinion

RedingJeff commented on examples/semantic/semantics_vocabulary_example.py:
```diff
@@ -0,0 +1,256 @@
+    print("")
+
+    # Here are two entities missing:
+    # o 'http://www.w3.org/2006/time#TemporalEntity'
```
It is a small O. Just wanted to make it look nicer, and it seemed to me as the best available bullet point substitute

RedingJeff commented on filip/clients/ngsi_v2/cb.py:
```diff
@@ -1532,7 +1528,6 @@ def patch_entity(self,
 
 
-
 #    def get_subjects(self, object_entity_name: str, object_entity_type: str, subject_type=None):
```
This code is not from me, it was always there I just left it.

To me it seems as updated subscription code, but I am not sure as I am not the author

RedingJeff commented on filip/semantics/vocabulary/combined_relation.py:
```diff
@@ -0,0 +1,91 @@
+    """user settings for field: Does this field needs to fulfilled to deem the 
+    instance fulfilled. Default: TRUE """
+
+    def get_relations(self, vocabulary: 'Vocabulary') -> List[Relation]:
```
Mostly due to legacy reasons.
I thought once about renaming them, but I liked the fact that they are, by having a slightly different name, better separable from Relationship Objects used in the Filip.

RedingJeff commented on tests/utils/test_simple_ql.py:
```diff
@@ -6,6 +6,7 @@
 
 
 class TestContextBroker(unittest.TestCase):
+
```
Test is not from me (only apperead in changelog, because I once added a skip statement)
I would add a comment, but I am honestly a bit unsure what it tests exactly and why it is named as it is

RedingJeff commented on filip/semantics/ontology_parser/rdfparser.py:
```diff
@@ -0,0 +1,712 @@
+    """Current source which is parsed, used for Log entries"""
+    current_class_iri = None
+    """Iri of class which is currently parsed, used for Log entries"""
+
```
Added now an init to make current class variables instances variables

RedingJeff commented on filip/semantics/ontology_parser/vocabulary_builder.py:
```diff
@@ -0,0 +1,206 @@
+
+class VocabularyBuilder(BaseModel):
+
+    vocabulary: Vocabulary
```
These should be instance variables as they are defined in a pydantic BaseModel or not?

RedingJeff commented on filip/semantics/semantic_manager.py:
```diff
@@ -0,0 +1,949 @@
+
+class SemanticManager(BaseModel):
+    """
+    The Semantic Manager is a static unique object that is delivered with
```
No i think not, the unique keyword is not correct. 
As your settings object it is only really used by the developer by importing the one instance that is predefined in the module.

RedingJeff commented on filip/semantics/ontology_parser/post_processer.py:
```diff
@@ -0,0 +1,634 @@
+        ontology.treat_dependency_statements(voc_builder.vocabulary)
+
+
+def add_predefined_datatypes(voc_builder: VocabularyBuilder):
```
For architectural reasons they are separated. But I now made Postprocesser as a class

RedingJeff commented on filip/semantics/ontology_parser/post_processer.py:
```diff
@@ -0,0 +1,634 @@
+        Datatype(iri="http://www.w3.org/2001/XMLSchema#hexBinary",
+                 comment="Hexadecimal",
+                 type=DatatypeType.string,
+                 allowed_chars=["0", "1", "2", "3", "4", "5", "6", "7", "8",
```
Nicely spotted!

RedingJeff commented on filip/semantics/ontology_parser/post_processer.py:
```diff
@@ -0,0 +1,634 @@
+            datatype.label = to_pascal_case(datatype.label)
+
+
+def save_initial_label_summary(vocabulary: Vocabulary):
```
No, it is a process of postprosessing that is only executed here. The user should later not see the method, as he should not use it.

RedingJeff commented on filip/semantics/semantic_manager.py:
```diff
@@ -0,0 +1,949 @@
+        """
+        self.instance_registry.load(json, self)
+
+    def visualise_local_state(self):
```
Ok, changed to american

RedingJeff commented:
Everything requested should be addressed or answered why it is designed in that way.

I remodeled the vocabulary files a bit, to have fewer small files, that maybe is a bit of a hindrance when backchecking the requests, but I think it was overall a good decision.

Everywhere it was possible I changed to Pydantic fields, and added all missing docs that I could find.

The most interesseting change that you may want to look at is the generated documentation for vocabulary models. I added the source ontology, as well as the comments extracted from the ontology entities.

tstorek commented on tests/.env:
```diff
@@ -0,0 +1,8 @@
+IOTA_JSON_URL="http://localhost:4041"
+IOTA_UL_URL="http://localhost:4061"
+QL_URL="http://localhost:8668"
+MQTT_BROKER_URL="mqtt://134.130.166.184:1883"
```
please remove IP

tstorek reviewed (changes requested):
almost there

RedingJeff commented:
I can not comment some of the comments, therefore here all answers:

- I added the LogLevel Model to models\base.py and used it for the tests and semantics
- Hmm IdType already got a docstring, maybe a visual error in git?
- I would strongly argue to stick with relation as a relation describes both relationships (object properties) and data properties.
- Could you concretise the last comment: "Still open?"

tstorek reviewed (changes requested):
one last thing

tstorek reviewed (changes requested):
please also rename tests and examples

RedingJeff commented:
> please also rename tests and examples

I renamed the package names to match: "semantics".

Should I also rename the file names? 
"test_semantic_models" sounds more correct than "test_semantics_models", and the tested model file is named "semantic_models". Same for the examples files.

tstorek commented:
@RedingJeff yes please

tstorek reviewed (approved):
@RedingJeff here we go! Thank you very much for this contribution :)
